### ISSUE TITLE
自定义神经元参数不更新

### ISSUE THREAD

**Issue by Taotaonoangry** [fangwei123456/spikingjelly#397]
我试着在自定义神经元中加入了一个可学习参数a，参考前面的问题，打印出过程，发现梯度并不更新，不管是继承LIFNode还是BaseNode，或者是修改学习率和优化器，a的值都不变，始终是初值0。希望能得到您的解答，万分感谢！

class TestLIFNode(neuron.BaseNode):
    def __init__(self, tau: float = 2., decay_input: bool = True, v_threshold: float = 1.,
                 v_reset: float = 0., surrogate_function: Callable = surrogate.Sigmoid(),
                 detach_reset: bool = False, step_mode='s', backend='torch', store_v_seq: bool = False, init_a : float = 0.):
        super().__init__(v_threshold, v_reset,surrogate_function, detach_reset,step_mode, backend, store_v_seq)

        self.decay_input = decay_input
        self.a = nn.Parameter(torch.as_tensor(init_a))
        self.tau = tau

    def neuronal_charge(self, x: torch.Tensor):
        self.v = self.v + (x - self.v) / self.tau + self.a * self.v

x_seq = torch.rand([8, 4], requires_grad=True)

net = TestLIFNode(surrogate_function=surrogate.ATan(), decay_input=True)
print(net)
print('a =', net.a)
optimier = torch.optim.SGD(net.parameters(), lr=0.1)

net(x_seq).sum().backward()

print(net.a.grad)
print('a.grad =', net.a)

optimier.step()

print('after bp, a =', net.a)

输出结果为：
TestLIFNode(
  v_threshold=1.0, v_reset=0.0, detach_reset=False, step_mode=s, backend=torch
  (surrogate_function): ATan(alpha=2.0, spiking=True)
)
a = Parameter containing:
tensor(0., requires_grad=True)
tensor(0.)
a.grad = Parameter containing:
tensor(0., requires_grad=True)
after bp, a = Parameter containing:
tensor(0., requires_grad=True)

**Comment by Yanqi-Chen** [comment]
Hi！从neuron_charge的式子来看，单步模式下，Loss对a的梯度是Loss对charge后v的梯度dL/dv1乘上charge前v的值dv1/da=v0，如果初始化v_reset=0，那么后者一定为0，所以Loss对a的梯度在这种情况下无论输入均恒为0，因此任何优化器都不会更新a。

**Comment by Taotaonoangry** [comment]
> Hi！从neuron_charge的式子来看，单步模式下，Loss对a的梯度是Loss对charge后v的梯度dL/dv1乘上charge前v的值dv1/da=v0，如果初始化v_reset=0，那么后者一定为0，所以Loss对a的梯度在这种情况下无论输入均恒为0，因此任何优化器都不会更新a。

感谢您的及时回复，问题解决了！非常感谢 (*^_^*)

**Comment by Zihao0** [comment]
> > Hi！从neuron_charge的式子来看，单步模式下，Loss对a的梯度是Loss对charge后v的梯度dL/dv1乘上charge前v的值dv1/da=v0，如果初始化v_reset=0，那么后者一定为0，所以Loss对a的梯度在这种情况下无论输入均恒为0，因此任何优化器都不会更新a。
> 
> 感谢您的及时回复，问题解决了！非常感谢 (_^_^_)

我也遇到类似的问题了，请问是怎么解决的啊？

**Comment by Zihao0** [comment]
> Hi！从neuron_charge的式子来看，单步模式下，Loss对a的梯度是Loss对charge后v的梯度dL/dv1乘上charge前v的值dv1/da=v0，如果初始化v_reset=0，那么后者一定为0，所以Loss对a的梯度在这种情况下无论输入均恒为0，因此任何优化器都不会更新a。

请问这种情况只能多步传播吗？且T必须大于1？

**Comment by Taotaonoangry** [comment]
> > > Hi！从neuron_charge的式子来看，单步模式下，Loss对a的梯度是Loss对charge后v的梯度dL/dv1乘上charge前v的值dv1/da=v0，如果初始化v_reset=0，那么后者一定为0，所以Loss对a的梯度在这种情况下无论输入均恒为0，因此任何优化器都不会更新a。
> > 
> > 
> > 感谢您的及时回复，问题解决了！非常感谢 (_^_^_)
> 
> 我也遇到类似的问题了，请问是怎么解决的啊？

给V_reset一个不为0的初值就可以了

**Comment by Zihao0** [comment]
> > > > Hi！从neuron_charge的式子来看，单步模式下，Loss对a的梯度是Loss对charge后v的梯度dL/dv1乘上charge前v的值dv1/da=v0，如果初始化v_reset=0，那么后者一定为0，所以Loss对a的梯度在这种情况下无论输入均恒为0，因此任何优化器都不会更新a。
> > > 
> > > 
> > > 感谢您的及时回复，问题解决了！非常感谢 (_^_^_)
> > 
> > 
> > 我也遇到类似的问题了，请问是怎么解决的啊？
> 
> 给V_reset一个不为0的初值就可以了

好的，非常感谢！
